### ISSUE TITLE
Game overrides don't work consistently

### ISSUE THREAD

**Issue by LimitCrown** [mgba-emu/mgba#1537]
I use the standalone version of mGBA. While trying to enable the GBA mode in The Legend of Zelda: Oracle of Seasons through two ways, I noticed that neither work correctly.

If I try to set the Game Boy Color model from Autodetect to Game Boy Advance in the Tools > Settings menu, then what happens is that GBA mode isn't triggered and instead the colors are dimmed, with for example the color white changing from rgb(255, 255, 255) to rgb(240, 240, 240). The colors appear to be scaled incorrectly. Changing this setting back to Autodetect makes it return to normal.

However, if I instead change the setting in Tools > Game overrides, GBA mode is enabled in the game the first time, but closing the game and running it again causes the above issue and the mode doesn't seem to be triggered. Unlike the other case, the option appears to be reverted back to Autodetect, but trying to save that option doesn't work and the colors remain muted until I clear my settings. I haven't checked what exactly changes in the files after changing the settings yet.

**Comment by endrift** [comment]
I don't seem to be able to reproduce on latest dev.

**Comment by LimitCrown** [comment]
Hmm. I downloaded the newest build of mGBA, but I'm running into the same issues still. If it helps, I've been using the .7z archive one and the x64 build. This is the muting of the colors that I'm referring to that occurs when I change the Game Boy Color model from Autodetect to Game Boy Advance in Tools > Settings:
![mGBA_2019-09-28_23-51-26](https://user-images.githubusercontent.com/24856516/66077199-7e7e3480-e52d-11e9-825c-92d2c47fde76.png)

When trying to change the game override setting, the config file does initially change to add the line and the game does recognize that it's being run on a GBA the first time that it's run, but afterwards it doesn't and the muting of colors occurs instead. Changing the game override setting back to autodetect doesn't actually remove the line from the config file, so I need to manually remove it to make the colors go back to their ordinary appearance.

**Comment by endrift** [comment]
I can reproduce the muting (it is indeed incorrect color scaling--mGBA uses a different color scaling for AGB mode but it doesn't scale up to 255), but I can't reproduce the Advance Shop not being open at the same time as the color muting being present.

**Comment by LimitCrown** [comment]
Ah. It seems it does make the Advance Shop still appear correctly, although the game still looks muted. The muting of the colors had confused me a bit and I should have checked every relevant location beforehand.

**Comment by LimitCrown** [comment]
Trying out the newest developmental build of mGBA, the colors initially appeared to be scaled correctly now at first when I used the option in Tools > Settings, but I now notice another discrepancy that may have been occurring along with the previous muting of color. When I compare the pictures that I take, the colors appear to be overall darker than they normally should.
The first picture is how the game should look when being run in GBA mode and the second is how it does when repeating the steps in my first post for finding the first error:

![mGBA_2019-10-03_11-12-20](https://user-images.githubusercontent.com/24856516/66141094-2862cd00-e5d1-11e9-9169-665b7f39282b.png)
![mGBA_2019-10-03_11-08-36](https://user-images.githubusercontent.com/24856516/66141115-2bf65400-e5d1-11e9-88de-92238d3950d7.png)

Using the game overrides option makes the game appear correctly when it's first run, but not when it's run in subsequent attempts similarly to the first error until the line in the config file is manually removed and the option is used again. My guess is that the emulator is performing its own gamma correction, which causes this discrepancy. However, I usually use color emulation shaders and this messes up the results.  Speaking of which, the gba-color shader was updated some time ago and there is a gbc-color one as well, so I've needed to edit the provided one for it to match.

**Comment by bentley** [comment]
Oracle of Ages and Oracle of Seasons themselves detect at runtime which system they’re on and select different palettes for Game Boy Color and Game Boy Advance.

**Comment by LimitCrown** [comment]
Yes, but the last two pictures are of the GBA palette being used. The first picture is with the values being scaled how they'd be expected to and the second is how mGBA currently displays them because of the different scaling that it does with Game Boy Advance being selected as the Game Boy Color model. The ordinary GBC palette looks different from the last two pictures and looks like this:
![mGBA_2019-09-28_12-12-38](https://user-images.githubusercontent.com/24856516/66222729-d50b8000-e69f-11e9-9790-9b0830a9b075.png)

The colors are scaled the normal way when GBC games are played in GBC mode, so it's only when a GBC game is played in GBA mode in mGBA where there is different scaling.

**Comment by RunisDauphin** [comment]
If you find it useful, you can use these romhacks to enable the GBA mode:
Oracle of Ages: http://www.romhacking.net/hacks/3580/
Oracle of Seasons: http://www.romhacking.net/hacks/3583/

Runis Dauphin
